src/server/game/Pools/PoolMgr.cpp

```diff
@@ -349,32 +320,70 @@ void PoolGroup<T>::SpawnObject(ActivePoolData& spawns, uint32 limit, uint32 trig
     if (triggerFrom)
         ++count;
 
-    // This will try to spawn the rest of pool, not guaranteed
-    for (int i = 0; i < count; ++i)
+    if (count > 0)
     {
-        PoolObject* obj = RollOne(spawns, triggerFrom);
-        if (!obj)
-            continue;
-        if (obj->guid == lastDespawned)
-            continue;
+        PoolObjectList rolledObjects;
+        rolledObjects.reserve(count);
 
-        if (obj->guid == triggerFrom)
+        // roll objects to be spawned
+        if (!ExplicitlyChanced.empty())
         {
-            ReSpawn1Object(obj);
-            triggerFrom = 0;
-            continue;
+            while (count && ExplicitlyChanced.size() > rolledObjects.size())
+            {
+                --count;
+                float roll = (float)rand_chance();
+
+                for (auto obj : ExplicitlyChanced)
```

Comment:
Two things here - first, you are creating a copy of `PoolObject` every time (no ref on type) and second, don't use auto if the type is not `iterator` (also avoids the first mistake you made since it makes you think about type you are using)

Reply:
TIL don't rely on compiler optimizations too much. I guess auto& would also work, I just like it more than writing types when I don't have to.

Reply:
This here isnt really about compiler optimizations btw but TC style guide - no auto except iterators (or other long nested types)